lib/Reader/readerir.cpp

```diff
@@ -1320,10 +1334,70 @@ GenIR::getType(CorInfoType CorType, CORINFO_CLASS_HANDLE ClassHandle,
   }
 }
 
+#define FEATURE_AVX_SUPPORT 1
+
+bool canUseAVX() 
+{
+#ifdef FEATURE_AVX_SUPPORT
+  return 1;
+#else
+  return false;
+#endif
+}
+
+#define XMM_REGSIZE_BYTES        16      // XMM register size in bytes
+#define YMM_REGSIZE_BYTES        32      // YMM register size in bytes
+
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+#define _TARGET_XARCH_
+#endif
+
+// Whether SSE2 is available
+bool canUseSSE2()
+{
+#ifdef _TARGET_XARCH_
+  return 1;//opts.compCanUseSSE2;
+#else
+  return false;
+#endif
+}
+
+unsigned getSIMDVectorRegisterByteLength()
+{
+#if defined(_TARGET_XARCH_) && !defined(LEGACY_BACKEND)
+  if (canUseAVX())
+  {
+    return YMM_REGSIZE_BYTES;
+  }
+  else
+  {
+    assert(canUseSSE2());
+    return XMM_REGSIZE_BYTES;
+  }
+#else
+  assert(!"getSIMDVectorRegisterByteLength() unimplemented on target arch");
+  unreached();
+#endif
+}
```

Comment:
It looks like most of the code above is from RyuJIT. I think that we'll probably want to do something that uses whatever target information LLVM can give us. Until we figure out how to do that, however (the best starting place for this is probably the feature bitset returned by `llvm::MCSubTargetInfo::getFeatureBits()`), I'd prefer that we clean this code up to remove the RyuJIT-isms. Perhaps something like:

```
bool canUseAVX() {
  // TODO: determine this based on the subfeatures of the target at runtime
  return true;
}

bool canUseSSE1() {
  // TODO: determine this based on the subfeatures of the target at runtime
  return true;
}

unsigned getSIMDVectorRegisterByteLength(Module& M) {
  Triple TargetTriple(M.getTargetTriple());

  switch (TargetTriple.getArch()) {
  case Triple::X86_64:
    const unsigned YMM
    if (canUseAVX()) {
      const unsigned YMMRegSizeInBytes = 32;
      return YMMRegSizeInBytes;
    } else {
      const unsigned XMMRegSizeInBytes = 16;
      return XMMRegSizeInBytes;
    }

  default:
    llvm_unreachable("Unsupported architecture");
  }
}
```